Requesting GET_ACCOUNTS permission on startup. #407

lemonboston opened the pull request:
@dmfs I am submitting this now so you can check and decide on UX steps and the exact messages, but the code is not finalized here. I will get back to this when the permission handling library is ready. (Or let me know if it's better to not wait for that.)

dmfs commented on opentasks/build.gradle:
```diff
@@ -49,5 +49,6 @@ dependencies {
     compile project(':opentasks-provider')
     compile 'com.google.android.apps.dashclock:dashclock-api:2.0.0'
     compile 'com.github.dmfs:color-picker:1.0'
+    compile 'com.github.dmfs.jems:iterators:1.7.1'
```
Just use `org.dmfs:iterators:1.7.1`. There is no need to specify the github jems version. In fact, the maven dependency name will remain the same even with the jems version. Same goes for `optional`.

lemonboston replied:
Ok, updated.

dmfs commented:
Ok. It's a start, but we need to tweak it here and there. Here are some initial notes:

* The description of the permission request is not clear or even wrong. It sounds as if OpenTasks would sync tasks which is not the case. Also there should be a link to the contacts permission issue.
    
    I'll work on a better text.
* If the UI finds any synced lists, it should not offer the option to not grant the permission. I.e. if we know that the user has synced lists, the permission is no longer optional. Just check if there are any task lists with an `account_type` other than the local type.

dmfs commented:
"don't show again" only means don't show again as long as there are no synced lists. If there is at least one synced list, the dialog should pop up again if it has been dismissed initially.
Also, the dialog should be modal. It should not be possible to dismiss it by taping on the background.

lemonboston commented:
I see, I have two questions.
OpenTaskPal would be useful for querying the task lists, it's not that important, but if this task is not urgent, can we wait until the tests task is complete so I could use OpenTaskPal here?
Would it be okay to start to use RxJava2 here for that query? I think it could be beneficial in general to start to use Rx for loadings with that kind of declarative async access approach we chatted about.

TODOs:

- [ ] Create and publish permissions library (Marten)
- [ ] Integrate permission library, possibly improve related design here (Gabor)
- [ ] Better dialog description text (Marten)
- [ ] Don't offer 'don't show again' button when there is any synced list (Gabor)
- [ ] Show dialog again when 'don't show again' had been selected, but synced list found (Gabor)
- [ ] Dialog should not be dismissible by tapping on background (Gabor)

dmfs commented:
@lemonboston please rebase on master

lemonboston commented:
Rebased on master.

My idea is to leave this on hold until permission lib is available because that can be a main driver of the code structure here. But let me know if I should continue working on the parts which are not closely related nevertheless.

dmfs commented:
@lemonboston please rebase again.

lemonboston commented:
Rebased and checked the workings again. Apart from the messages it looks good to me as a first approximation.
If user denies with "don't show again" at the permission grant dialog, but not at ours, then after tapping OK on ours, nothing happens. But that's kind of an edge case.

Also updated the dialog to not be dismissable by tapping outside.

dmfs commented:
Please replace the dialog text with
```
In order to read the list of accounts on your device which support task synchronization, OpenTasks needs the GET_ACCOUNTS permission.

Please note that (due to the great wisdom of the Android developers) the permission request will be labeled as "access your contacts" instead of "read your accounts". See <a href="https://issuetracker.google.com/issues/37065963">Issue 37065963</a> for further information.
``'

lemonboston commented:
Updated the dialog, ready for check again.

lemonboston commented:
As requested, I've remove the opt out option for granting the permission. It's forced on the user now, checked in `BaseActivity.onResume()`. I've also added handling the 'don't ask again' case, for that the button changes to `OPEN SETTINGS` and opens the phone settings for the app when clicked.
That could have a slightly different message, but I guess that's not important.

dmfs commented:
@lemonboston please rebase.

lemonboston commented:
Rebased.

dmfs commented:
Merging this despite the CI issue. It works for me.
